### PR TITLE
Plutus core/fix typechecker

### PR DESCRIPTION
**Pull request by vmchale**
Fixes the typechecker. I think it also introduces a performance regression, which I will put in a separate ticket.

### CODE REVIEW

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 1]
```diff
@@ -228,43 +231,54 @@ typeOf (TyInst x t ty) = do
...
-    if tyReduce fixed == ty'
+    red <- tyReduce fixed
+    if red == ty'
         then pure (TyFix () (void n) (void ty))
         else throwError (TypeMismatch x (void t) fixed (void ty'))
 
 extractUnique :: TyNameWithKind a -> Unique
 extractUnique = nameUnique . unTyName . unTyNameWithKind
 
+fixUniversals :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind a)
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: What does this do? Why do we need it? Can we not do this in a more principled way?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 2]
```diff
@@ -228,43 +231,54 @@ typeOf (TyInst x t ty) = do
...
                 else throwError (KindMismatch x (void ty) k k')
         _ -> throwError (TypeMismatch x (void t) (TyForall () dummyTyName dummyKind dummyType) (void ty'))
 typeOf (Unwrap x t) = do
     ty <- typeOf t
     case ty of
         TyFix _ n ty' -> do
-            let subst = tySubstitute (extractUnique n) ty ty'
-            pure (tyReduce subst)
+            subst <- tySubstitute (extractUnique n) ty ty'
+            typeCheckStep
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Somewhat unrelated, but I'm confused about what the policy is about where these go.

File: language-plutus-core/test/types/recursiveUnwrap.plc [hunk 3]
```diff
@@ -1,18 +0,0 @@
-(program 1.0.0
```
Comment (michaelpj) [inline language-plutus-core/test/types/recursiveUnwrap.plc]: Why did you delete this test?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 4]
```diff
@@ -96,7 +101,7 @@ defaultTable = do
     bsRelType <- bsRel
 
     let f = M.fromList .* zip
-        termTable = f intTypes is <> f intRelTypes irs <> f [TxHash, EqByteString] [txHash, bsRelType]
+        termTable = f intTypes is <> f intRelTypes irs <> f [TxHash, EqByteString ] [txHash, bsRelType ]
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Why the strange formatting?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 5]
```diff
@@ -54,7 +59,7 @@ bsRel = builtinRel TyByteString
 -- | Create a dummy 'TyName'
 newTyName :: (MonadQuote m) => Kind () -> m (TyNameWithKind ())
 newTyName k = do
-    u <- nameUnique . unTyName <$> liftQuote (freshTyName () "a")
+    u <- liftQuote freshUnique
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Unrelated, but perhaps we should generalize functions like `freshUnique` to work for any `MonadQuote m`?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 6]
```diff
@@ -113,13 +118,12 @@ kindCheck :: (MonadError (Error a) m, MonadQuote m) => Natural -> Type TyNameWit
 kindCheck n t = convertErrors asError $ runTypeCheckM n (kindOf t)
 
 -- | Run the type checker with a default context.
-runTypeCheckM :: (MonadError (TypeError a) m, MonadQuote m)
-              => Natural -- ^ Amount of gas to provide typechecker
+runTypeCheckM :: Natural -- ^ Amount of gas to provide typechecker
               -> TypeCheckM a b
-              -> m b
+              -> ExceptT (TypeError a) Quote b
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: I do not have any particular preference here, but I'm curious why you changes the signature.

**Comment by mchakravarty** [comment]
If you are mentioning a deficiency (such as a performance regression), please always describe what it is and why you think it is a problem. Going through the changes, trying to guess what you are referring to is tedious.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 1 above)
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Yes, probably. I wanted something Chad could use in the meantime, mainly.

File: language-plutus-core/test/types/recursiveUnwrap.plc (hunk 3 above)
Comment (vmchale) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: It's not valid according to the spec.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 6 above)
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: This was just what worked. If we need to modify it, that can be done.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 1 above)
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: As for how it works: currently, when substituting in a value (which Chad noticed with `unwrap`), we substitute naivlely. If, however, we substitute a term containing universals, this can mean that the global uniqueness of names is not preserved, because we didn't generate a new bound variable.

**Comment by vmchale** [comment]
The performance deficiency (not sure if I should document this elsewhere) is that this rewrites every term every time we substitute a value. The substitutions are already sub-optimal, so this is quite unfortunate. 

The difference is indeed noticeable; the typechecker benchmarks now take longer.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 2 above)
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: I believe @effectfully suggested that I  use `Language.PlutusCore.Constant.Typed`.

File: language-plutus-core/test/types/recursiveUnwrap.plc (hunk 3 above)
Comment (michaelpj) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: Why not fix it? Reducing our test coverage is bad.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: I mean the `typeCheckStep`s? Are we talking about the same thing?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 1 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Please put that in the code.

On the code itself:
- What about type lambdas and fixpoints? Those both bind variables and so will suffer from the same problem when substituting.
- Perhaps it would be clearer to have a function called something like `freshenBoundVariables` that replaces all bound variables with fresh names? That's the general case, right?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 6 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: I don't see what was wrong with the old type signature either.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 1 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Also, I think we could probably construct some reduced test cases that illustrate precisely the issue, rather than being as large as Chad's example.

File: language-plutus-core/test/types/fold.plc.golden [hunk 7]
```diff
@@ -0,0 +1 @@
+Error at 10:19. Variable a is not in scope.
```
Comment (michaelpj) [inline language-plutus-core/test/types/fold.plc.golden]: Is this supposed to happen?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 6 above)
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Nothing is *wrong* per se, however it will not compile as currently written with the old type signature.

File: language-plutus-core/test/types/recursiveUnwrap.plc (hunk 3 above)
Comment (vmchale) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: It's not obvious to me how to do so, but I will have a look at fixing it.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 6 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Well you also changed the body...

File: language-plutus-core/test/types/fold.plc.golden (hunk 7 above)
Comment (vmchale) [reply language-plutus-core/test/types/fold.plc.golden]: Nope. Shoot. I'll change this to WIP.

File: language-plutus-core/test/types/recursiveUnwrap.plc (hunk 3 above)
Comment (michaelpj) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: Looks valid to me - what's wrong with it?

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 2 above)
Comment (vmchale) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Oh, no I was confused. I believe they accompany a `tySubstitute`. Perhaps I should put them before?

File: language-plutus-core/test/types/fold.plc.golden (hunk 7 above)
Comment (vmchale) [reply language-plutus-core/test/types/fold.plc.golden]: Oh, perhaps this test case is in error. The `a` that is bound is a type variable. I will ping Chad.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: I don't know - hence why I asked what the policy was! I think we could do with a note somewhere in this module about what things we consider to consume gas - but that can be done separately.

File: language-plutus-core/test/types/recursiveUnwrap.plc (hunk 3 above)
Comment (vmchale) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: This is confusing, since it actually involves an old version of the spec. In the spec this repo is currently tied to, a `(lam x ty t)` must have `ty` be a type value. Thus, the 

```
(fix list (all r (type) (fun r (fun (fun [(con integer) (con 64)] (fun list r)) r))))
```

type annotation is not valid.

Comment (michaelpj) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: Hm, I see. In the current version there's a workaround to avoid the fact that we can't reduce builtin applications for terms - the special "builtin value" rule allows us to have unreduced builtin applications in values. Sadly we're missing the corresponding thing at the type level. AFAIC that means that it's impossible to abstract over types containing builtin types atm. Good thing we're getting rid of type values!

However, an easy fix suggests itself for the test: simply parameterize the function by a type and a value to use in the function. That should then work fine. It's a bit more verbose than the current version using concrete types, but oh well.

File: language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs [hunk 8]
```diff
@@ -211,64 +215,81 @@ typeOf (Constant _ (BuiltinName _ n)) = do
 typeOf (Constant _ (BuiltinInt _ n _))           = pure (integerType n)
 typeOf (Constant _ (BuiltinBS _ n _))            = pure (bsType n)
 typeOf (Constant _ (BuiltinSize _ n))            = pure (sizeType n)
-typeOf (Apply x fun arg) = do
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Could you not rename all the arguments back again? That was deliberate because it makes it a lot clearer. `ty'''` is gibberish.

Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs]: Yes please.

File: language-plutus-core/test/types/recursiveUnwrap.plc (hunk 3 above)
Comment (vmchale) [reply language-plutus-core/test/types/recursiveUnwrap.plc]: That seems sensible to me. I will try to do this.
